source/server/listener_manager_impl.cc

```diff
@@ -239,8 +244,72 @@ ListenerImpl::~ListenerImpl() {
   filter_factories_.clear();
 }
 
-bool ListenerImpl::createNetworkFilterChain(Network::Connection& connection) {
-  return Configuration::FilterChainUtility::buildFilterChain(connection, filter_factories_);
+bool ListenerImpl::isWildcardServerName(const std::string& name) {
+  return name.size() > 2 && name[0] == '*' && name[1] == '.';
+}
+
+void ListenerImpl::addFilterChain(const std::string& transport_socket_name,
+                                  const std::vector<std::string>& server_names,
+                                  Network::TransportSocketFactoryPtr&& transport_socket_factory,
+                                  std::vector<Network::NetworkFilterFactoryCb> filters_factory) {
+  ASSERT(transport_socket_factory != nullptr);
+  ASSERT(!filters_factory.empty());
+
+  const auto filter_chain = std::make_shared<FilterChainImpl>(std::move(transport_socket_factory),
+                                                              std::move(filters_factory));
+  // Save mappings.
+  if (server_names.empty()) {
+    filter_chain_exact_match_factories_[transport_socket_name][EMPTY_STRING] = filter_chain;
+  } else {
+    for (const auto& name : server_names) {
+      if (isWildcardServerName(name)) {
+        filter_chain_wildcard_match_factories_[transport_socket_name][name] = filter_chain;
+      } else {
+        filter_chain_exact_match_factories_[transport_socket_name][name] = filter_chain;
+      }
+    }
+  }
+}
+
+const Network::FilterChainSharedPtr
+ListenerImpl::findFilterChain(const std::string& transport_protocol_name,
```

Comment:
The documentation says "The FilterChain with the most specific FilterChainMatch criteria is used on a connection."

I think this is too imprecise to be useful. What exactly does "most specific" mean? Given that every field except sni_domains is marked not-implemented-hide, I think this algorithm is fine. But as soon as we un-hide any of the other fields, we'll have to revisit this, and update the documentation to specify how it works.

FWIW, this is the matching algorithm that I want/need for my use case: hash-table lookup by SNI, as opposed to O(n) walking a list looking for matches.

Reply:
Yes, we'll need to revisit this sooner rather than later (i.e. I need to add filter chain matching based on the destination IP:port to get ride of the `bind_to_port` setting), but I think we can iterate on it as we unhide features, instead of blocking this PR on that.